### ISSUE TITLE
XFeat + LightGlue

### ISSUE THREAD

**Issue by guipotje** [verlab/accelerated_features#34]
Hello everyone,

I'm training some LightGlue variations (finding a neat trade-off between model size vs accuracy) and soon I will update the repo with the model and weights in the next weeks!

You can follow this issue if you are interested.

Best,

Guilherme

**Comment by guipotje** [comment]
Hey guys, just released a version of LightGlue matcher, please check out in [README ](https://github.com/verlab/accelerated_features?tab=readme-ov-file#xfeat--lightglue)

**Comment by muchluv525** [comment]
> 大家好，刚刚发布了一个版本的 LightGlue 匹配器，请在 [README ](https://github.com/verlab/accelerated_features?tab=readme-ov-file#xfeat--lightglue) 中查看

Hello, author, didn't you say that XFEAT performance is better than Superpoint, why look at the data this time Superpoint is stronger

**Comment by guipotje** [comment]
Hi @muchluv525,

Please notice that we trained a *smaller and faster* version of LightGlue.

Other than that, there are still a few reasons that SuperPoint + LightGlue might be still better than XFeat + LG (full size)
1 - Descriptor embedding: XFeat extracts a much more compact (64-D vs 256-D) descriptors;
2 - SuperPoint has a much larger backbone.

**Comment by zhouzq-thu** [comment]
And the number of layers is just 6. BTW, would you upload the training code of LightGlue for XFeat?

**Comment by muchluv525** [comment]
> 嗨，
> 
> 请注意，我们训练了一个_更小、更快_的 LightGlue 版本。
> 
> 除此之外，SuperPoint + LightGlue 可能仍然优于 XFeat + LG（全尺寸）有几个原因 1 - 描述符嵌入：XFeat 提取更紧凑（64-D 与 256-D）的描述符; 2 - SuperPoint 具有更大的主干网。
I see, thank you for the answer, I'm a beginner and look forward to a more detailed explanation of xfeat

**Comment by muchluv525** [comment]
> 嗨，
> 
> 请注意，我们训练了一个_更小、更快_的 LightGlue 版本。
> 
> 除此之外，SuperPoint + LightGlue 可能仍然优于 XFeat + LG（全尺寸）有几个原因 1 - 描述符嵌入：XFeat 提取更紧凑（64-D 与 256-D）的描述符; 2 - SuperPoint 具有更大的主干网。

我明白了，谢谢你的回答，我是初学者，期待对xfeat进行更详细的解释

**Comment by noahzn** [comment]
@guipotje Have you tried training Xfeat and LightGlue end-to-end?

**Comment by guipotje** [comment]
Hello @noahzn, I haven't tried to train it end-to-end. It might deliver some improvements, as mentioned in SuperGlue's paper (section 5.4), when backpropagating through the descriptors. However, it also might lead to less generalization to different scenes.

**Comment by zhangchang0127** [comment]
Hello, I encountered an error while converting the ONNX model using torch.export. Here is the code we used for the model conversion. 
![2024-08-12 14-43-37 的屏幕截图](https://github.com/user-attachments/assets/e540feda-b102-41ec-a94a-a022f1511c6c)

This is the error.
![2024-08-12 14-44-29 的屏幕截图](https://github.com/user-attachments/assets/44cc7046-b300-4e0d-bdf6-3098a1eea5ca)

**Comment by EndlessPeak** [comment]
Hello,

Thank you very much for your great work. 

I am curious about the modifications made to the lightglue network structure to achieve the balance between inference accuracy and speed mentioned in the README. Will this part of the code be made publicly available?

I checked the match_lighterglue.py file, but it does not provide more information on this aspect.

**Comment by noahzn** [comment]
> Hello @noahzn, I haven't tried to train it end-to-end. It might deliver some improvements, as mentioned in SuperGlue's paper (section 5.4), when backpropagating through the descriptors. However, it also might lead to less generalization to different scenes.

Hi @guipotje I might try end-to-end training, do you have any idea about the implementation? Since XFeat and LightGlue use different homography code for training, I'm wondering if it's possible to keep their own random homography code, but only optimize two networks together. This may need XFeat to return a loss and then we add it to LightGlue's loss and backprop together. Do you think this is the minimal effort for an end-to-end training of the two networks?

**Comment by noahzhy** [comment]
Great work! I notice that lighter version of your lightglue which is selected `n_layers=6`, as we know the original paper of lightglue with some metrics when L=3, 5, 7, 9, I am a little curious about your `n_layers` settings.

**Comment by zw-92** [comment]
Hello, thank you very much for your work. However, when I evaluated the notebooks' xfeat + lightglue on Megadepth1500, I found that sometimes the output matching points were too small, causing the set verification (RANSAC) to fail. Do you have any suggestions? Looking forward to your reply.

**Comment by July0928** [comment]
hi，we find when we set topk=4096, the score of some keypoints<0, finally we get the number of keypoints is<4096, then more batch is wrong, how can we get the same numbers of keypoints?

**Comment by hhl134** [comment]
Hi, may I ask if xfeat+lightglue involves complete pre-training (revisitop1m) and fine-tuning (megadepth)? When I was doing pre-training, lightglue could not converge, and the number of matching points was also very small. Is there a complete training process explanation?  thanks!
